### ISSUE TITLE
Research kratix -> crossplane migration

### ISSUE THREAD

**Issue by piontec** [giantswarm/roadmap#3739]
# Lessons learned, comparison and recommendations for using Crossplane and Kratix

- [Lessons learned, comparison and recommendations for using Crossplane and Kratix](#lessons-learned-comparison-and-recommendations-for-using-crossplane-and-kratix)
  - [Introduction](#introduction)
  - [TL;DR and Recommendations](#tldr-and-recommendations)
  - [Analysis](#analysis)
    - [How do they work?](#how-do-they-work)
    - [How are they similar?](#how-are-they-similar)
    - [How are they different?](#how-are-they-different)
  - [Notes](#notes)
  

## Introduction

We wanted to compare Kratix and Crossplane and decide which one to use to continue our IDP demo effort. As an
experiment, we have re-implemented some existing Kratix based features in Crossplane. This document summarizes
our experience and provides recommendations for using these tools.

On a very high level, Kratix and Crossplane do the same thing: they allow you to quickly create new CRDs that
are then "executed", which usually means rendering a set of different related resources as a result. That way
one should be able to provide custom workflows and abstractions that build your IDP API.

## TL;DR and Recommendations

It is hard to pick just one, as they differ a lot. Simplified, Kratix is like bash, Crossplane is like go.
Which one you should use depends on your use case and your team's skills. Only Crossplane allows to manage
infrastructure, so if that's your goal, Crossplane is the choice. Still, keep in mind that one can combine
Kratix and Crossplane in multiple ways, as they are not mutually exclusive.

Taking into account functionality and maturity, if you have to choose just one, choose Crossplane, but be
prepared for a steep learning curve.

### How do they work?

Kratix uses the term "Promise" to describe an abstraction unit. A Promise is basically a CRD + a definition of
a set of containers that are executed in a certain order (called Pipeline). Containers in a pipeline get the
CR (a request for an instance) as YAML file and have to render resulting desired objects as files in an output
directory. The result (the files in the output dir) are then committed to a git repository. Kratix doesn't
apply them to a k8s cluster, you have to provide a tool like Flux to take manifest files from the git
repository and apply them to a cluster.

Crossplane is much more complex. It separates the API definition and its implementation. The API is included
in an object called XRD (Crossplane Resource Definition, very similar to plain CRDs). Then, for each XRD,
there can be multiple "Compositions", which act on it and produce requested effects (like groups of other
objects to create). Compositions are created by defining "pipelines", which describe a sequence of dedicated
"Crossplane Functions" to be invoked on the requested resource. Each "function" is really a gRPC server that
takes the requested object on input and renders more objects on output of a gRPC remote call. In the end,
output is applied to k8s API server. There are many functions already provided, and in theory you should be
able to do most of what you need with them. Still, sometimes you need to create your own function and there's
a go and python SDK for that.

### How are they similar?

- both take a CR and allow you to act on it, with the main use case of creating other (Custom) Resources; in
  other words, they allow you to abstract sets of inter-connected resources into a single object.
- both have a concept of a pipeline, which is a set of containers that are executed in a certain order to
  provide the output set of Resources

### How are they different?

> Note: this is a very high-level comparison, and there are many more differences between the two projects.
> I'm not comparing infrastructure management abilities, as Kratix doesn't have them. The part I'm interested
> in is using the tools to build IDPs in general.

- Building blocks:
  - Kratix allows you to use whatever-and-however-container as a step in a pipeline, as long as it saves some
    output files in an output directory (Kratix provides the requested input resource as a YAML file in a
    directory and expects the output to be saved as files in another directory). An API object in Kratix is
    called a "Promise" and its instance is called a "Request".
  - Crossplane requires you to compose your logic from containers running dedicated gRPC functions. The basic
    abstraction is a "Composition", which is a sequence of "Functions" that are executed on the input object
    (called either XR or Claim), which API (spec) is defined in an XRD.
- Platform composability:
  - In Kratix, it's close to none. If you want to combine or reuse existing promises, you have to provide a
    full CRD/API definition, then provide code that will render out existing sub-promises as output files the
    way you need them. There's no way to "include" a promise in another promise without code duplication.
  - Crossplane is more "composable" than Kratix: it separates the API (XRD) from implementation (Composition),
    allows you to connect the API to different implementations.
    - Additionally, there are many composition functions that provide tools that make defining Composistions
      and their output objects (composition's result) easier. They include Functions for tools like KCL or
      Cue, which enable code reuse, schemas, and types verification for objects generated in a composition
      run.
    - For API definitions (XRD), there's a crossbuilder tool (maintained now by Martin in his fork) that
      allows you to specify and compose the API using the same mechanism as in kubebuilder, using go lang
      code. Crossbuilder also allows for better composition and management of Compositions, as they tend to be
      huge files and this allows to build them from snippets in go code.
- Dependency management: rather limited in both.
  - Kratix: for a given Promise you can list other promises that need to be already available when the new
    Promise is installed.
  - Crossplane: you can list functions and providers your composition requires.
- Versioning:
  - In crossplane, each Composition, when changed, creates a CompositionRevision (immutable); XRs (created
    resources) either automatically reference the newest Revision or can opt for manual Revision reference
  - Kratix has a concept of [PromiseReleases](https://docs.kratix.io/main/reference/promises/releases) but it
    seems (I haven't tested) that they can only be used to ensure that a single Promise is installed in one
    specific version (version lock). It can't be used to provide different versions of the same Promise at the
    same time. Additionally, Kratix runs a set of hooks whenever a Promise definition is changed, but it's up
    to the user to handle potential migration and changes in the related Requests.
- Debugging:
  - Crossplane is harder to debug, as in practice you want to run your composition to get a reliable output.
    You can try and run it locally, but that proves to be hard because of all the dependencies, including
    potentially building and uploading your own functions to a registry. However, for the function development
    itself, it generates a unit test stub for you, that allows you to test the function in isolation (calls to
    the gRPC server). There's also an option to run your gRPS Function container with Docker locally, and
    render out an output, but it's not that practical.
  - In Kratix, you can write your code to run in the container in whatever language, using this language's
    testing tools. Then, to test the container (one step in the pipeline), it's pretty easy, as you can just
    do "specific files in -> expected files out" testing. Testing whole pipelines is however also problematic
    and best done on a real cluster.
- Maturity: Crossplane is more mature, with a lot of providers and a lot of features; Kratix is still very
  much in development. It's also important to note big gap between the communities and companies behind the
  projects, where crossplane's is much bigger.
- Security:
  - Crossplane: each provider and each Composition Function run in a separate pods and can have a
    ServiceAccount configured
    - Specific, but important: `provider-kubernetes`, in its default config at least, gives effectively
      `clustr-admin` to anyone having permissions to create crossplane objects.
  - Kratix: allows to set securityContext and SerciceAccount for the whole pipeline
    ([docs](https://docs.kratix.io/main/reference/workflows#rbac)). Still, Kratix tends to run extra
    containers in each pipeline pod, which tend to be hard to configured (we were hit by a problem where you
    couldn't set the securityContext for the extra containers in a previous versions).
- Barrier to entry: Kratix is much easier to get started with, as you can just write a container in any
  language and run it in the pipeline. Crossplane is much more complex, with even the basic set of XRD + XR +
  Claim + Composition being hard to understand and get right. Adding more Composition Functions, Providers and
  their configs makes it even harder. If you want to use crossbuilder to keep your code DRY, enable better
  composability and strict type checking, this grows even further.
- Unique features:
  - Kratix:
    - Each pipeline can actually have sub-pipelines, each sub-pipeline is applied separately through the
      "config delivery", in this case Flux. In theory allows for pausing or gating the creation process or to
      better manage output objects dependencies (i.e. it can wait with an app deployment until the db is
      ready).
    - It doesn't apply and rendered resources to a cluster, actually it doesn't even have a k8s client. It
      saves the resulting state into a "State Store", like a GitOps repository, and it's up to the user to
      apply them.
    - It is pretty easy to target multiple k8s clusters with Kratix, as you can write output files to multiple
      locations in the gitops repo and then deliver them to the correct cluster with the gitops tool (Kratix
      doesn't apply them anywhere, it just saves the output files to a git repo).
  - Crossplane:
    - It has much more features, like accidental resource deletion protection, ability to provide multiple
      ProviderConfigs that use different credentials and related permissions.

## Notes

- don't use `function-shell`, it's a disaster
  - super hard to debug - the gRPC server that runs shell code has problems with returning reliable output
    when a script execution fails
  - all script invocations share the same long-lived pod to execute all the requested scripts - beware of
    persistent state
- in Crossplane, there's a Context passed between each function invocation step in a Composition. Many
  functions use that to pass some information not possible to contain in the XR or requested composed
  resources. We use that as well to pass a GitHub Token from our
  [GitHub App Auth Function](https://github.com/giantswarm/function-github-app-get-token) to all subsequent
  functions in a Composition.

**Comment by uvegla** [comment]
The `AppDeployment` crossplane resources and examples are here: https://github.com/giantswarm/crossplane-gs-apis/pull/4/files

**Comment by uvegla** [comment]
The sketch for shell script using GS IDP function shell to template the generated repository is here: https://github.com/giantswarm/dev-platform-crossplane/tree/github-repo-xr/

**Comment by mproffitt** [comment]
In general I agree with the outcome of the discussion here. It's fair and on point.

One thing to note, with Crossplane, depedencies are not just limited to functions and providers, but may also include other compositions packaged as `xpkg` images. This way, you can enforce that the installation of one composition package results in the installation of all dependant compositions.

The drawback to this approach now lies with how crossplane is managing the providers and the inability to specify provider source as part of the package definition which effectively breaks the packaging spec at this present time.

It may be possible to change this at the cluster level by aliasing `xpkg.upbound.io` to our `gsoci` but that has not been explored and for the moment, packages must be installed without dependency resolution
